Make System.Guid readonly

GrabYourPitchforks opened the pull request:
This is a follow-up to https://github.com/dotnet/runtime/pull/1809. It contains two changes: (a) make `System.Guid` readonly (+ some minor code cleanup); and (b) expose `Unsafe.Unbox<T>` for internal use, now that `Guid` can use it as an implementation detail. :)

One piece of feedback given in https://github.com/dotnet/runtime/pull/1809 was that we shouldn't use the internal `MutableGuid` type. I've changed the call sites here to use `Unsafe.AsRef`, but to be honest I think this is introducing noise and that `MutableGuid` is a cleaner solution. If we don't like `MutableGuid`, then perhaps we can clone all of the GUID fields into the existing mutable type `GuidResult`, then introduce an API `GuidResult.ToGuid()` which is basically a glorified _reinterpret\_cast_.

I _think_ I hit all the needed places to shim `Unsafe.Unbox`. From investigating other calls, as long as we have a rudimentary implementation in _Unsafe.cs_ we don't need to enlighten the mono interpreter.

GrabYourPitchforks committed:
Make System.Guid readonly
- Also adds Unsafe.Unbox for internal use

jkotas commented:
> we can clone all of the GUID fields into the existing mutable type GuidResult

I like this idea. GuidResult can actually have just 4 ints (instead of the bytes and shorts from Guid). It may make the parsing tiny bit faster - we won't need to split the parser int into bytes.

jkotas commented on src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/Unsafe.cs:
```diff
@@ -420,5 +420,22 @@ public static void SkipInit<T>(out T value)
+        /// <summary>
+        /// Returns a mutable reference to the contents of a boxed struct. The caller should take
+        /// care not to reassign the value except through the struct's own mutable API surface.
+        /// The type of the boxed object is not checked.
```
> /// The type of the boxed object is not checked.

It is not checked in this impl, but it is checked in the intrinsic impl...

GrabYourPitchforks replied:
I've updated the comment to be more accurate. I don't think it's all that important to have the two implementations be kept completely in sync, as this is an internal-only API and we can reasonably expect our callers to follow basic hygiene. Do you agree that it's ok to have the two different behaviors here?

jkotas replied:
The idea behind the Internal S.R.CS.Unsafe is that it behaves the same way as the public one.

I do not think that the performance of `Guid.Equals(object o)` really matters. I think it would be ok to keep it as is. If it gets a bit slower as the side-effect of the change, I would not care much. My guess is that the current version of the change makes it slower anyway because of there will be an extra call or jump compared to the baseline.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -769,33 +769,33 @@ public bool TryWriteBytes(Span<byte> destination)
+
+            // Compare each element
+
+            if (rA != rB) { return false; }
```
Is this code pattern working around JIT limitations?

GrabYourPitchforks replied:
The same reason that `string.IsNullOrEmpty` has that weird _? true : false_ construct at the end. But honestly, since this method isn't getting inlined anywhere, it's really not critical to optimize this. I'll simplify.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -769,33 +769,33 @@ public bool TryWriteBytes(Span<byte> destination)
+            // Compare each element
+
+            if (rA != rB) { return false; }
+            if (Unsafe.Add(ref rA, 1) != Unsafe.Add(ref rB, 1)) { return false; }
```
You can also change this to do the comparison as longs using `Unsafe.ReadUnaligned`. We had abandoned PR for that: https://github.com/dotnet/runtime/pull/35654

GrabYourPitchforks replied:
Do you know why it was abandoned? I see the alignment discussion on that thread, but there doesn't seem to be any clear reason the PR was rejected.

jkotas replied:
I guess Bruce had more important things to do than to finish it.

EgorBo replied:
Something like this I guess 🙂 
```csharp
private static bool EqualsCore_1(in Guid left, in Guid right)
{
    ref byte leftBytes = ref Unsafe.As<int, byte>(ref Unsafe.AsRef(in left._a));
    ref byte rightBytes = ref Unsafe.As<int, byte>(ref Unsafe.AsRef(in right._a));

    // early out if not equal
    if (Unsafe.ReadUnaligned<ulong>(ref leftBytes) != 
        Unsafe.ReadUnaligned<ulong>(ref rightBytes))
        return false;

    return Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref leftBytes, 8)) == 
           Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref rightBytes, 8));
}
```
a SIMD version of it slightly faster for equal guids but in most cases, I guess, we can "early out"

GrabYourPitchforks replied:
Using 64-bit comparisons didn't seem to help matters - see https://github.com/dotnet/runtime/pull/44629#issuecomment-727076381.

jkotas replied:
|              Equal_32Bit | 0.9002 ns | 0.0087 ns | 0.0082 ns |
|              Equal_64Bit | 1.1797 ns | 0.0109 ns | 0.0102 ns |

Interesting. https://github.com/dotnet/runtime/pull/35654 had this case as an improvement.

benaadams commented:
Also needs the `AsRef` for assignments in Guid.Unix.cs?

GrabYourPitchforks committed:
Update Guid.Unix.cs

GrabYourPitchforks commented:
Here's the perf run from `Guid.Equals(in Guid a, in Guid b)` my machine, where the equality operation was using 32-bit comparisons, 64-bit comparisons, branchless 64-bit comparisons, and 128-bit comparisons (SIMD). This assumes that the equality operator function __is not__ being inlined. The branch predictor has also been primed.

```txt
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.51904, CoreFX 5.0.20.51904), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.51904, CoreFX 5.0.20.51904), X64 RyuJIT


|                   Method |      Mean |     Error |    StdDev |
|------------------------- |----------:|----------:|----------:|
|              Equal_32Bit | 0.9002 ns | 0.0087 ns | 0.0082 ns |
|              Equal_64Bit | 1.1797 ns | 0.0109 ns | 0.0102 ns |
|    Equal_64BitBranchless | 0.8910 ns | 0.0054 ns | 0.0050 ns |
|               Equal_Sse2 | 0.9096 ns | 0.0080 ns | 0.0075 ns |
|           NotEqual_32Bit | 1.1562 ns | 0.0082 ns | 0.0073 ns |
|           NotEqual_64Bit | 1.1845 ns | 0.0091 ns | 0.0085 ns |
| NotEqual_64BitBranchless | 0.8972 ns | 0.0091 ns | 0.0086 ns |
|            NotEqual_Sse2 | 0.9080 ns | 0.0069 ns | 0.0058 ns |
```

I'm not seeing any measurable improvement in using 64-bit comparisons over 32-bit comparisons. The branchless / SIMD versions can be a little better, but now we're operating in the picosecond range. Doesn't seem worth the complexity to me.

GrabYourPitchforks commented:
On the plus side, I experimented with replacing `Guid.GetHashCode()` with one-round AES, since a 128-bit Guid is perfectly sized to mimic one block of plaintext. And this knocked the `GetHashCode` runtime down to 0.25 nanoseconds on my box. 'Twas a fun experiment, and now the Crypto Board will stare daggers at me for the rest of my days.

GrabYourPitchforks committed:
PR feedback: Make GuidResult mutable

GrabYourPitchforks committed:
Fix remarks in Unsafe.cs

GrabYourPitchforks commented:
With latest iteration, I tried a hybrid of the "move all the fields to `GuidResult`" and "is it possible to make `GuidResult` contain only 4 ints?" approaches. This hybrid approach uses `[FieldOffset]` so that both the "parse D" and "parse X" logic feel somewhat natural and don't end up polluted with width-changing `Unsafe.As` calls everywhere.

If this approach isn't preferred, we can investigate alternatives.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -373,33 +402,29 @@ private static bool TryParseExactD(ReadOnlySpan<char> guidString, ref GuidResult
-                        g._d = (byte)(uintTmp >> 8);
-                        g._e = (byte)uintTmp;
+                        // _d, _e must be stored as a big-endian ushort
+                        result._de = (BitConverter.IsLittleEndian) ? BinaryPrimitives.ReverseEndianness((ushort)uintTmp) : (ushort)uintTmp;
```
```suggestion
                        result._de = BitConverter.IsLittleEndian ? BinaryPrimitives.ReverseEndianness((ushort)uintTmp) : (ushort)uintTmp;
```

Nit - multiple places. I believe the usual style for src/libraries is without the extra `( )` in this context.

`(BitConverter.IsLittleEndian) ?` - 27 occurences under src/libraris
`BitConverter.IsLittleEndian ?` - 3 occurrences under src/libraries

GrabYourPitchforks replied:
Maybe this is an argument to add `BinaryPrimitives.htonl`, which would encapsulate the check and make the call sites cleaner? Right now I think this API sits on `IPAddress`, which is the wrong layering.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -529,7 +551,7 @@ private static bool TryParseExactX(ReadOnlySpan<char> guidString, ref GuidResult
 
             // Read in the number
-            if (!TryParseHex(guidString.Slice(numStart, numLen), out result._parsedGuid._b, ref overflow) || overflow)
+            if (!TryParseHex(guidString.Slice(numStart, numLen), out Unsafe.As<ushort, short>(ref result._b), ref overflow) || overflow)
```
Change the type of `_b` to short to avoid this `Unsafe.As` ?

jkotas replied:
Or change the argument of TryParseHex to ushort.

jkotas commented:
> This hybrid approach uses [FieldOffset] so that both the "parse D" and "parse X" logic feel somewhat natural and don't end up polluted with width-changing Unsafe.As calls everywhere

I like this hybrid approach.

GrabYourPitchforks committed:
PR feedback & remove Unsafe.Unbox

GrabYourPitchforks commented:
Latest iteration: removed internal `Unsafe.Unbox` implementation, removed `Unsafe.As` wherever I could by re-typing the internal APIs, code cleanup w.r.t. unnecessary parens.

Crossing my fingers and hoping CI cooperates today. :)

GrabYourPitchforks commented:
Thanks @jkotas for the feedback. Much appreciated! :)

GrabYourPitchforks commented:
/azp run runtime
